File: apps/risk-service/src/infrastructure/messaging/outbox_publisher_worker.py
import logging
import threading
from typing import Optional

from sqlalchemy import text

from infrastructure.messaging.kafka_producer import KafkaProducerWrapper
from infrastructure.persistence.db import SessionLocal

logger = logging.getLogger(__name__)

MAX_RETRIES = 10
BATCH_SIZE = 100
POLL_INTERVAL_SECONDS = 1.0

_SELECT_PENDING_SQL = text(
    """
    SELECT id, aggregate_id, event_type, payload, retry_count
    FROM outbox_events
    WHERE status = 'PENDING' AND retry_count < :max_retries
    ORDER BY created_at
    LIMIT :batch_size
    FOR UPDATE SKIP LOCKED
    """
)

_MARK_PUBLISHED_SQL = text(
    "UPDATE outbox_events SET status = 'PUBLISHED', published_at = now(), error = NULL WHERE id = :id"
)

_MARK_RETRY_SQL = text(
    "UPDATE outbox_events SET status = :status, retry_count = :retry_count, error = :error WHERE id = :id"
)


class OutboxPublisherWorker:
    """Mismo patrón que ledger-service (Java) y los workers TypeScript: polling con
    SELECT ... FOR UPDATE SKIP LOCKED, batch=100, maxRetries=10, marca
    PENDING/PUBLISHED/FAILED. Corre en un hilo daemon separado del event loop de FastAPI."""

    def __init__(self, producer: KafkaProducerWrapper, topic: str) -> None:
        self._producer = producer
        self._topic = topic
        self._stop_event = threading.Event()
        self._thread: Optional[threading.Thread] = None

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, daemon=True, name="outbox-publisher-worker")
        self._thread.start()

    def stop(self) -> None:
        self._stop_event.set()
        if self._thread:
            self._thread.join(timeout=5)

    def _run(self) -> None:
        while not self._stop_event.is_set():
            try:
                self.publish_pending_events()
            except Exception:
                logger.exception("Outbox publisher tick failed")
            self._stop_event.wait(POLL_INTERVAL_SECONDS)

    def publish_pending_events(self) -> None:
        with SessionLocal() as session:
            rows = (
                session.execute(_SELECT_PENDING_SQL, {"max_retries": MAX_RETRIES, "batch_size": BATCH_SIZE})
                .mappings()
                .all()
            )

            for row in rows:
                try:
                    self._producer.send(self._topic, key=row["aggregate_id"], value=row["payload"])
                    session.execute(_MARK_PUBLISHED_SQL, {"id": row["id"]})
                except Exception as exc:
                    retry_count = row["retry_count"] + 1
                    status = "FAILED" if retry_count >= MAX_RETRIES else "PENDING"
                    session.execute(
                        _MARK_RETRY_SQL,
                        {"status": status, "retry_count": retry_count, "error": str(exc)[:1000], "id": row["id"]},
                    )

            session.commit()

File: apps/customer-service/src/infrastructure/messaging/outbox-publisher.worker.ts
import { Injectable, Logger } from '@nestjs/common';
import { Interval } from '@nestjs/schedule';
import { ConfigService } from '@nestjs/config';
import { DataSource } from 'typeorm';
import { KafkaEventPublisher } from './kafka-event-publisher';
import { withEventTraceContext } from '../observability/trace-context.util';

interface PendingOutboxRow {
  id: string;
  aggregate_id: string;
  event_type: string;
  payload: string;
  retry_count: number;
}

const MAX_RETRIES = 10;
const BATCH_SIZE = 100;
const POLL_INTERVAL_MS = 1000;

@Injectable()
export class OutboxPublisherWorker {
  private readonly logger = new Logger(OutboxPublisherWorker.name);

  constructor(
    private readonly dataSource: DataSource,
    private readonly producer: KafkaEventPublisher,
    private readonly config: ConfigService,
  ) {}

  @Interval(POLL_INTERVAL_MS)
  async publishPendingEvents(): Promise<void> {
    const topic =
      this.config.get<string>('redpanda.topicCustomerEvents') || 'banking.customer.events';
    const queryRunner = this.dataSource.createQueryRunner();
    await queryRunner.connect();
    await queryRunner.startTransaction();

    try {
      const pending: PendingOutboxRow[] = await queryRunner.manager.query(
        `SELECT id, aggregate_id, event_type, payload, retry_count
         FROM outbox_events
         WHERE status = 'PENDING' AND retry_count < $1
         ORDER BY created_at
         LIMIT $2
         FOR UPDATE SKIP LOCKED`,
        [MAX_RETRIES, BATCH_SIZE],
      );

      for (const row of pending) {
        try {
          const envelope = JSON.parse(row.payload) as {
            correlationId?: string;
            causationId?: string;
          };
          await withEventTraceContext(envelope.correlationId, envelope.causationId, () =>
            this.producer.sendRaw(topic, row.aggregate_id, row.payload),
          );
          await queryRunner.manager.query(
            `UPDATE outbox_events SET status = 'PUBLISHED', published_at = now(), error = NULL WHERE id = $1`,
            [row.id],
          );
        } catch (error) {
          const retryCount = row.retry_count + 1;
          const status = retryCount >= MAX_RETRIES ? 'FAILED' : 'PENDING';
          const message = (error instanceof Error ? error.message : String(error)).slice(0, 1000);
          await queryRunner.manager.query(
            `UPDATE outbox_events SET status = $1, retry_count = $2, error = $3 WHERE id = $4`,
            [status, retryCount, message, row.id],
          );
        }
      }

      await queryRunner.commitTransaction();
    } catch (error) {
      await queryRunner.rollbackTransaction();
      this.logger.error({ err: error }, 'Outbox publisher tick failed');
    } finally {
      await queryRunner.release();
    }
  }
}
